src/main/java/com/google/collegeplanner/servlets/DatastoreServlet.java

```diff
@@ -0,0 +1,179 @@
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.apache.http.client.utils.URIBuilder;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+
+/** Queries the UMD API and downloads the data to datastore. */
+@WebServlet("/api/download")
+public class DatastoreServlet extends BaseServlet {
+  final int PAGE_LIMIT = 200;
+
+  DatastoreService datastore;
+
+  public DatastoreServlet() {
+    this(DatastoreServiceFactory.getDatastoreService(), new ApiUtil());
+  }
+
+  public DatastoreServlet(DatastoreService datastore, ApiUtil apiUtil) {
+    super(apiUtil);
+    this.datastore = datastore;
+  }
+
+  /*
+   * Downloads all course data to datastore.
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Loop through the pages.
+    // We don't know how many pages there are beforehand. We stop looping when the new page
+    // doesn't have any results.
+    int page = 1;
+    do {
+      URI uri;
+      try {
+        URIBuilder builder = new URIBuilder("https://api.umd.io/v1/courses");
+        builder.setParameter("page", Integer.toString(page));
+        uri = builder.build();
+      } catch (URISyntaxException e) {
+        respondWithError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, response);
+        return;
+      }
+      JSONArray coursesArray = apiUtil.getJsonArray(uri);
+      if (coursesArray == null) {
+        respondWithError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, response);
+        return;
+      } else if (coursesArray.size() == 0) {
+        // Successful. The default status code response is 200.
+        return;
+      }
+      addCourses(coursesArray, response);
+    } while (page++ < PAGE_LIMIT);
+  }
+
+  /**
+   * Creates a Course object from a JSONArray.
+   * @param coursesArray The course json from the UMD API.
+   * @param response The HttpServletResponse object.
+   */
+  private void addCourses(JSONArray coursesArray, HttpServletResponse response) throws IOException {
+    if (coursesArray == null) {
+      return;
+    }
+
+    // Loop through each page's courses.
+    for (Object jsonObject : coursesArray) {
+      JSONObject courseJson = (JSONObject) jsonObject;
+
+      // Create a Course object from the JSONObject for parsing.
+      Course course;
+      try {
+        course = new Course(courseJson);
+      } catch (Exception e) {
+        respondWithError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, response);
+        return;
+      }
+
+      // Check if the entity already exists in datastore, and modify it if it does.
+      Query query = new Query("Course").setFilter(
```

Comment:
side question - is there a way in datastore to enforce uniqueness on an attribute like `course_id`, so we'll never accidentally commit two at the same time (if there are two scripts running simultaneously for example).

Reply:
Enforce it with a datastore setting? I don't believe so. The way I did it queries the actual database anyways, so if one script is running and it adds the entity slightly before the other script does, then this query should pick up that the entity is already in the database, right?

Reply:
Concurrent writes are difficult to get right, especially in distributed systems. There's always a chance that both writes will try to happen simultaneously.

Also, the concurrent scripts likely won't hit the same server, so it's not always possible to pick up an "already-written" entity in the database due to small sync/propagation delays.

Reply:
Through my research online I haven't been able to find a better solution. If we needed to seriously handle lots of concurrent writes then we would've had to use Firestore or something similar.

Reply:
It looks like we can solve this by just using the course_id as the key. You would trade off correctness with any potential key distribution problems - seems like a worthwhile tradeoff for our use case here to me.